File: Freedom/主界面/LibraryCollectionViewController.m
//  LibraryCollectionViewController.m
//  Freedom
//  Created by Super on 16/6/24.
//  Copyright © 2016年 Super. All rights reserved.
#import "LibraryCollectionViewController.h"
#import "User.h"
@interface LibraryCollectionViewCell : BaseCollectionViewCell
@end
@implementation LibraryCollectionViewCell
-(void)initUI{
    [super initUI];
    self.icon.frame = CGRectMake(0, 0, APPW/5, 80);
    self.icon.layer.cornerRadius = 40;
    self.icon.clipsToBounds = YES;
    self.title.frame = CGRectMake(0, YH(self.icon), W(self.icon), 20);
    self.title.textAlignment = NSTextAlignmentCenter;
    self.backgroundColor = [UIColor clearColor];
    [self addSubviews:self.icon,self.title,nil];
}
-(void)setCollectionDataWithDic:(NSDictionary *)dict{
    self.title.text = [dict valueForKey:@"title"];
    self.icon.image = [UIImage imageNamed:[dict valueForKey:@"icon"]];
}
@end
@interface LibraryCollectionViewController (){
    NSMutableArray *libraryBooks;
}
@end
@implementation LibraryCollectionViewController
@synthesize contentLength, dismissByBackgroundDrag, dismissByBackgroundTouch, dismissByForegroundDrag;
-(id)initWithCoder:(NSCoder *)aDecoder{
    self = [super initWithCoder:aDecoder];
    if (self) {
        self.contentLength = APPW;
        self.dismissByBackgroundTouch   = YES;
        self.dismissByBackgroundDrag    = YES;
        self.dismissByForegroundDrag    = YES;
    }return self;
}
static NSString * const reuseIdentifier = @"Cell";
- (void)viewDidLoad {
    [super viewDidLoad];
    UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc]init];
    layout.itemSize = CGSizeMake(APPW/5, 100);
    layout.minimumInteritemSpacing = 10;
    layout.scrollDirection = UICollectionViewScrollDirectionVertical;
    layout.minimumLineSpacing = 30;
    layout.sectionInset = UIEdgeInsetsMake(30, 10, 0, 10);
    self.collectionView = [[UICollectionView alloc]initWithFrame:CGRectMake(0, 0, APPW, APPH-110) collectionViewLayout:layout];
    self.collectionView.backgroundColor = whitecolor;
    UIImageView *backView = [[UIImageView alloc]initWithFrame:CGRectMake(0, 100, APPW, APPH-100)];
    backView.image = [UIImage imageNamed:@""];
    self.collectionView.backgroundView = backView;
    ElasticTransition *ET = (ElasticTransition*)self.transitioningDelegate;
    DLog(@"\ntransition.edge = %@\ntransition.transformType = %@\ntransition.sticky = %@\ntransition.showShadow = %@", [HelperFunctions typeToStringOfEdge:ET.edge], [ET transformTypeToString], ET.sticky ? @"YES" : @"NO", ET.showShadow ? @"YES" : @"NO");
    
    libraryBooks = [NSMutableArray arrayWithArray:[User getControllerData]];
    [self.collectionView registerClass:[LibraryCollectionViewCell class] forCellWithReuseIdentifier:reuseIdentifier];
    self.collectionView.delegate = self;
    self.collectionView.dataSource = self;
}
#pragma mark <UICollectionViewDataSource>
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return libraryBooks.count;
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    LibraryCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:reuseIdentifier forIndexPath:indexPath];
    if(!cell){
        cell = [[LibraryCollectionViewCell alloc]initWithFrame:CGRectMake(0, 0, APPW/5, 100)];
    }
    [cell setCollectionDataWithDic:libraryBooks[indexPath.item]];
    return cell;
}
-(void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath{
    NSString *controlName = [libraryBooks[indexPath.row] valueForKey:@"control"];
    [[UIApplication sharedApplication] setStatusBarHidden:NO];
    if(![controlName isEqualToString:@"Sina"]){
        UIStoryboard *StoryBoard = [UIStoryboard storyboardWithName:controlName bundle:nil];
        [self showViewController:[StoryBoard instantiateViewControllerWithIdentifier:[NSString stringWithFormat:@"%@TabBarController",controlName]] sender:self];
        return;
    }
    NSString *s =[NSString stringWithFormat:@"%@TabBarController",controlName];
    UIViewController *con = [[NSClassFromString(s) alloc]init];
    CATransition *animation = [CATransition animation];
    animation.duration = 1;
    animation.timingFunction = UIViewAnimationCurveEaseInOut;
    [self.view.window.layer addAnimation:animation forKey:nil];
    [self presentViewController:con animated:NO completion:^{
    }];
    [SVProgressHUD showSuccessWithStatus:[libraryBooks[indexPath.row] valueForKey:@"title"]];
}
#pragma mark <UICollectionViewDelegate>
//- (BOOL)collectionView:(UICollectionView *)collectionView shouldHighlightItemAtIndexPath:(NSIndexPath *)indexPath {
//	return YES;
//}
//- (BOOL)collectionView:(UICollectionView *)collectionView shouldSelectItemAtIndexPath:(NSIndexPath *)indexPath {
//    return YES;
//}
//- (BOOL)collectionView:(UICollectionView *)collectionView shouldShowMenuForItemAtIndexPath:(NSIndexPath *)indexPath {
//	return NO;
//}
//- (BOOL)collectionView:(UICollectionView *)collectionView canPerformAction:(SEL)action forItemAtIndexPath:(NSIndexPath *)indexPath withSender:(id)sender {
//	return NO;
//}
//- (void)collectionView:(UICollectionView *)collectionView performAction:(SEL)action forItemAtIndexPath:(NSIndexPath *)indexPath withSender:(id)sender {
//	
//}
//- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath{
//    return CGSizeMake(APPW/5, 80);
//}
@end

File: Freedom/主界面/LibraryCollectionViewController.swift
//
//  LibraryCollectionViewController.swift
//  Freedom
//
//  Created by Super on 6/15/18.
//  Copyright © 2018 Super. All rights reserved.
//
import UIKit
import XExtension
class LibraryCollectionViewCell: BaseCollectionViewCell {
    override func initUI() {
        super.initUI()
        icon.frame = CGRect(x: 0, y: 0, width: APPW / 5, height: 80)
        icon.layer.cornerRadius = 40
        icon.clipsToBounds = true
        title.frame = CGRect(x: 0, y: icon.bottom, width: icon.width, height: 20)
        title.textAlignment = .center
        backgroundColor = UIColor.clear
        addSubviews([icon, title])
    }
}
class LibraryCollectionViewController: BaseViewController,ElasticMenuTransitionDelegate,UICollectionViewDelegate,UICollectionViewDataSource {
    var collectionView: BaseCollectionView!
    var contentLength:CGFloat = APPW
    var items:[[String:String]] = (UIApplication.shared.delegate as! AppDelegate).items
    var dismissByBackgroundTouch = true
    var dismissByBackgroundDrag = true
    var dismissByForegroundDrag = true
    override func viewDidLoad() {
        super.viewDidLoad()
        let layout = UICollectionViewFlowLayout()
        layout.itemSize = CGSize(width: APPW / 5, height: 100)
        layout.minimumInteritemSpacing = 10
        layout.scrollDirection = .vertical
        layout.minimumLineSpacing = 30
        layout.sectionInset = UIEdgeInsets(top: 30, left: 10, bottom: 0, right: 10)
        collectionView = BaseCollectionView(frame: CGRect(x: 0, y: 0, width: APPW, height: APPH - 110), collectionViewLayout: layout)
        collectionView.backgroundColor = UIColor.white
        let backView = UIImageView(frame: CGRect(x: 0, y: 100, width: APPW, height: APPH - 100))
        backView.image = UIImage(named: "")
        collectionView?.backgroundView = backView
        let tm = transitioningDelegate as! ElasticTransition
        print( "transition.edge = .\(tm.edge)\n" +
            "transition.transformType = .\(tm.transformType)\n" +
            "transition.sticky = \(tm.sticky)\n" +
            "transition.showShadow = \(tm.showShadow)")
        collectionView.register(LibraryCollectionViewCell.self, forCellWithReuseIdentifier: LibraryCollectionViewCell.identifier)
        collectionView.delegate = self
        collectionView.dataSource = self
        view.addSubview(collectionView)
    }
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return items.count
    }
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        var cell = collectionView.dequeueCell(LibraryCollectionViewCell.self, for: indexPath)
        let dict = items[indexPath.row]
        cell.title.text = dict["title"]
        cell.icon.image = UIImage(named: dict["icon"]!)
        return cell
    }
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let mode = PopoutModel()
        mode.name = items[indexPath.row]["control"]!
        AppDelegate.radialView.didSelectBlock!(AppDelegate.radialView, false,false,mode)
    }
}
